test/models/binomial.py

```diff
@@ -0,0 +1,72 @@
+from typing import Union
+from scipy import stats
+from scipy.special import logit, expit as inv_logit, log1p
+import numpy as np
+import numpy.typing as npt
+
+
+class Binomial:
+    """
+    Binomial model with a (conjugate) beta prior.
+
+    This model is decomposed as a prior and a likelihood for
+    testing with samplers that require that, e.g. SMC. The
+    posterior has a closed-form beta distribution.
+    """
+
+    def __init__(
+        self,
+        alpha: float = 2,
+        beta: float = 3,
+        x: int = 5,
+        N: int = 15,
+        seed: Union[None, int, np.random.BitGenerator, np.random.Generator] = None,
+    ) -> None:
+        self.alpha = alpha
+        self.beta = beta
+        self.x = x
+        self.N = N
+        self._rand = np.random.default_rng(seed)
+
+    def dims(self) -> int:
+        return 1
+
+    def log_density(self, params_unc: npt.NDArray[np.float64]) -> float:
+        return self.log_likelihood(params_unc) + self.log_prior(params_unc)
+
+    def log_prior(self, theta: npt.NDArray[np.float64]) -> float:
+        theta_constrained = inv_logit(theta[0])
+        jac_adjust: float = np.log(theta_constrained) + log1p(-theta_constrained)
+        prior: float = stats.beta.logpdf(theta_constrained, self.alpha, self.beta)
+        return prior + jac_adjust
+
+    def log_likelihood(self, theta: npt.NDArray[np.float64]) -> float:
+        theta_constrained = inv_logit(theta[0])
+        return stats.binom.logpmf(self.x, self.N, theta_constrained)  # type: ignore # scipy is not typed
+
+    def initial_state(self, _: int) -> npt.NDArray[np.float64]:
+        return logit(self._rand.beta(self.alpha, self.beta, size=1))  # type: ignore # scipy is not typed
+
+    def log_density_gradient(
+        self, params_unc: npt.NDArray[np.float64]
+    ) -> tuple[float, npt.NDArray[np.float64]]:
+        # use finite diffs for now
```

Comment:
Finite diffs are suboptimal given low precision (here the speed is probalby close as this is univariate and not using central finite diffs).   Probably not a big deal, but it's also easy to work out the gradient using Wolfram alpha.  But isn't this going to change in the tempered setting anyway?

Reply:
We don't use the gradient in the tempered setting, this is only there for HMC and MALA to use.

I haven't been able to get WolframAlpha or matrix-calculus.org to spit out the derivative of this distribution. Maybe I'm just using the wrong keywords. It would necessarily involve taking derivatives of the gamma function though, so it is probably not worth it for code that is only used in testing